pilot/pkg/serviceregistry/federation/controller.go

```diff
@@ -0,0 +1,1116 @@
+	return c
+}
+
+// UpdatePeerConfig updates all settings derived from the ServiceMeshPeer resource.
+// It will also restart the watch if the controller is already running.
+func (c *Controller) UpdatePeerConfig(peerConfig *v1.ServiceMeshPeer) {
+	if func() bool {
+		c.storeLock.RLock()
+		defer c.storeLock.RUnlock()
+		return c.peerConfigGeneration == peerConfig.Generation
+	}() {
+		return
+	}
+
+	restartWatch := func() bool {
+		c.storeLock.Lock()
+		defer c.storeLock.Unlock()
+
+		restartWatch := false
+
+		c.discoveryServiceName = common.DiscoveryServiceHostname(peerConfig)
+		c.defaultDomainSuffix = defaultDomainSuffixForMesh(peerConfig)
+		c.namespace = peerConfig.Namespace
+		c.useDirectCalls = peerConfig.Spec.Security.AllowDirectOutbound
+		if c.egressName != peerConfig.Spec.Gateways.Egress.Name {
+			c.egressName = peerConfig.Spec.Gateways.Egress.Name
+			c.egressService = fmt.Sprintf("%s.%s.%s",
+				peerConfig.Spec.Gateways.Egress.Name, c.namespace, c.localDomainSuffix)
+			c.discoveryURL = fmt.Sprintf("%s://%s:%d", common.DiscoveryScheme, c.egressService, common.DefaultDiscoveryPort)
+			restartWatch = true
+		}
+
+		if !reflect.DeepEqual(c.remote, peerConfig.Spec.Remote) {
+			c.remote = peerConfig.Spec.Remote
+			restartWatch = true
+		}
+
+		c.peerConfigGeneration = peerConfig.Generation
+
+		return restartWatch
+	}()
+	if c.hasStarted() && restartWatch {
+		c.RestartWatch()
+	}
+}
+
+// UpdateImportConfig updates the import rules that are used to select services for
+// import into the mesh and rewrite their names.
+func (c *Controller) UpdateImportConfig(importConfig *v1.ImportedServiceSet) {
+	if func() bool {
+		if importConfig == nil {
+			return false
+		}
+		c.storeLock.RLock()
+		defer c.storeLock.RUnlock()
+		return c.importConfigGeneration == importConfig.Generation
+	}() {
+		return
+	}
+	func() {
+		c.storeLock.Lock()
+		defer c.storeLock.Unlock()
+		c.importNameMapper = common.NewServiceImporter(importConfig, nil, c.defaultDomainSuffix, c.localDomainSuffix)
+		if importConfig == nil {
+			c.importLocality = nil
+		} else {
+			c.importLocality = importConfig.Spec.Locality
+			c.importConfigGeneration = importConfig.Generation
+		}
+		c.locality = mergeLocality(c.importLocality, c.defaultLocality)
+	}()
+}
+
+func (c *Controller) Cluster() cluster.ID {
+	return c.clusterID
+}
+
+func (c *Controller) Provider() provider.ID {
+	return provider.Federation
+}
+
+func (c *Controller) pollServices() (*federationmodel.ServiceListMessage, error) {
+	url := c.discoveryURL + "/v1/services/"
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: '%s': %s", url, err)
+	}
+	req.Header.Add("discovery-service", c.discoveryServiceName)
+	req.Header.Add("remote", fmt.Sprint(common.RemoteChecksum(c.remote)))
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to GET URL: '%s': %s", url, err)
```

Comment:
WDYT about adding more context to this error message, like: `"failed to poll services [URL=%s]: %s", err`?